multiraft/multiraft.go

```diff
@@ -428,14 +436,18 @@ type state struct {
 	groups    map[proto.RangeID]*group
 	nodes     map[proto.RaftNodeID]*node
 	writeTask *writeTask
+	// buffer the event if s.Events channel is full to avoid the deadlock
+	// between s.Events channel and callbackChan.
+	pendingEvents []interface{}
```

Comment:
It will be a little faster to have the buffered channel, but how much faster? Most of the time the select loop should be idle, so it should be able to hand off the next event to the processing goroutine. I'm not convinced the channel buffer is worth the complexity, but I'm willing to be persuaded if you see a difference.